src/algorithms/xp.cpp

```diff
@@ -0,0 +1,427 @@
+        size_t end = rank == path_count ? pn_iv.size() : pn_bv_select(rank+1);
+        end -= 1;  // step before '$'
+        std::string name; name.resize(end-start);
+        for (size_t i = start; i < end; ++i) {
+            name[i-start] = pn_iv[i];
+        }
+        return name;
+    }
+
+    void XP::serialize_members(std::ostream& out) const {
+        serialize_and_measure(out);
+    }
+
+    size_t XP::serialize_and_measure(std::ostream& out, sdsl::structure_tree_node* s, std::string name) const {
+
+        sdsl::structure_tree_node* child = sdsl::structure_tree::add_child(s, name, sdsl::util::class_name(*this));
+        size_t written = 0;
+
+        // Do the magic number
+        out << "XP";
+        written += 2;
+
+        // POSITION MAP STUFF
+        written += sdsl::write_member(pos_map_iv, out, child, "position_map");
+        // PATH STUFF
+        written += sdsl::write_member(path_count, out, child, "path_count");
+
+        // Treat the paths as their own node
+        size_t paths_written = 0;
+        auto paths_child = sdsl::structure_tree::add_child(child, "paths", sdsl::util::class_name(*this));
+
+        paths_written += pn_iv.serialize(out, paths_child, "path_names");
+        paths_written += pn_csa.serialize(out, paths_child, "path_names_csa");
+        paths_written += pn_bv.serialize(out, paths_child, "path_names_starts");
+        paths_written += pn_bv_rank.serialize(out, paths_child, "path_names_starts_rank");
+        paths_written += pn_bv_select.serialize(out, paths_child, "path_names_starts_select");
+        paths_written += pi_iv.serialize(out, paths_child, "path_ids");
+
+        for (size_t i = 0; i < paths.size(); i++) {
+            XPPath* path = paths[i];
+            paths_written += path->serialize(out, paths_child, "path:" + XP::get_path_name(handlegraph::as_path_handle(i+1)));
+        }
+
+        sdsl::structure_tree::add_size(paths_child, paths_written);
+        written += paths_written;
+
+        sdsl::structure_tree::add_size(child, written);
+        return written;
+    }
+
+    void XP::deserialize_members(std::istream& in) {
+        // simple alias to match an external interface
+        load(in);
+    }
+
+    void XP::load(std::istream& in) {
+
+        if (!in.good()) {
+            throw XPFormatError("Index file does not exist or index stream cannot be read");
+        }
+
+        // We need to look for the magic value
+        // TODO @ekg is this necessary?!
+        char buffer;
+        in.get(buffer);
+        if (buffer == 'X') {
+            in.get(buffer);
+            if (buffer == 'P') {
+                // We found the magic value!
+
+            } else {
+                // Put back both characters
+                in.unget();
+                in.unget();
+            }
+        } else {
+            // Put back the one character
+            in.unget();
+        }
+
+        try {
+            sdsl::read_member(pos_map_iv, in);
```

Comment:
This and the following line are responsible for the problem. I just don't know why. @ekg

Reply:
So only `sdsl::read_member(pos_map_iv, in);` is at fault....... hmmmm

Reply:
So the problem was, when serializing the index to file, I had to serialize the pos_map_iv to disk instead of storing it as a member via sdsl.